x/expiration/types/expiration.go

```diff
@@ -0,0 +1,109 @@
+package types
+
+import (
+	"fmt"
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/cosmos/cosmos-sdk/codec/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+func NewExpiration(
+	moduleAssetID string,
+	owner string,
+	time time.Time,
+	deposit sdk.Coin,
+	message types.Any,
+) *Expiration {
+	return &Expiration{
+		ModuleAssetId: moduleAssetID,
+		Owner:         owner,
+		Time:          time,
+		Deposit:       deposit,
+		Message:       message,
+	}
+}
+
+// ValidateBasic basic format checking of the data
+func (e *Expiration) ValidateBasic() error {
+	now := time.Now()
+	if strings.TrimSpace(e.ModuleAssetId) == "" {
+		return ErrEmptyModuleAssetID
+	}
+	if err := validateAddress(e.ModuleAssetId); err != nil {
+		return err
+	}
+	if strings.TrimSpace(e.Owner) == "" {
+		return ErrEmptyOwnerAddress
+	}
+	if e.Time.Before(now) {
+		return ErrTimeInPast
+	}
+	if !e.Deposit.IsValid() {
+		return ErrInvalidDeposit
+	}
+	if err := e.validateMessage(); err != nil {
+		return ErrInvalidMessage
+	}
+	return nil
+}
+
+// validateMessage validates `expiration.Message` conforms to `sdk.Msg`
+// and is registered and whitelisted with the InterfaceRegistry
+func (e *Expiration) validateMessage() error {
+	// this may occur in message decoding
+	if e.Message.Value == nil {
+		return fmt.Errorf("expecting non nil value to validate an Any message")
+	}
+	if e.Message.TypeUrl == "" {
+		return fmt.Errorf("expecting message type URL to unpack Any")
+	}
+	// validate message is a whitelisted sdk.Msg during a BroadcastTx
+	var msg sdk.Msg
+	if err := ModuleCdc.UnpackAny(&e.Message, &msg); err != nil {
+		return err
+	}
+	if msg == nil {
+		return fmt.Errorf("failed to unpack Any message: %v", msg)
+	}
+	return msg.ValidateBasic()
+}
+
+var reDuration = regexp.MustCompile(`(^[1-9]\d{0,11})([ywdh])$`)
+
+// ParseDuration parses a duration string
+func ParseDuration(s string) (*time.Duration, error) {
+	// FindStringSubmatch returns a slice of strings holding the text of the
+	// leftmost match of the regular expression in s and the matches, if any.
+	matches := reDuration.FindStringSubmatch(s)
+	if len(matches) != 3 {
+		return nil, ErrDurationValue
+	}
+
+	// parse digits
+	digit, err := strconv.ParseInt(matches[1], 10, 64)
+	if err != nil {
+		return nil, err
+	}
+
+	period := matches[2]
+	var duration time.Duration
+	switch period {
+	case "y":
+		duration = time.Duration(digit) * 24 * 365 * time.Hour
+	case "w":
+		duration = time.Duration(digit) * 24 * 7 * time.Hour
+	case "d":
+		duration = time.Duration(digit) * 24 * time.Hour
+	case "h":
+		duration = time.Duration(digit) * time.Hour
```

Comment:
I don't think I saw anywhere documenting the duration strings. It should probably be documented that `1d` = `24h`, `1w` = `7d` = `168h`, `1y` = `365d` = `8760h`. There are several definitions of the word "year", and even "day". Also, that it's not like a `time.Duration` in that you can only provide one unit, e.g. you can't do `1y5w`.

Reply:
Please also document it in the proto. Spec documentation is good too, but the protos are the primary source of truth and the most readily available source of documentation that a consumer has.

Reply:
comment added to the proto:
```
// the duration period by which to extend the existing expiration period
// the accepted values for a duration are n{h,d,w,y}
// where 1h = 60m, 1d = 24h, 1w = 7d (or 168h), 1y = 365d (or 8760h)
string duration = 2;
  ```